n3fit/src/evolven3fit_new/eko_utils.py

```diff
@@ -58,46 +57,75 @@ def construct_eko_cards(
     theory = Loader().check_theoryID(theoryID).get_description()
     theory.pop("FNS")
     theory.update(theory_card_dict)
-    if "nfref" not in theory:
-        theory["nfref"] = NFREF_DEFAULT
-    if "nf0" not in theory:
-        theory["nf0"] = NF0_DEFAULT
-        
+
     # Prepare the thresholds according to MaxNfPdf
     thresholds = {"c": theory["kcThr"], "b": theory["kbThr"], "t": theory["ktThr"]}
     if theory["MaxNfPdf"] < 5:
         thresholds["b"] = np.inf
     if theory["MaxNfPdf"] < 6:
         thresholds["t"] = np.inf
-        
+
+    if "nfref" not in theory:
+        theory["nfref"] = NFREF_DEFAULT
+
+    # Set nf_0 according to the fitting scale unless set explicitly
+    mu0 = theory["Q0"]
+    if "nf0" not in theory:
+        if mu0 < theory["mc"] * thresholds["c"]:
+            theory["nf0"] = 3
+        elif mu0 < theory["mb"] * thresholds["b"]:
+            theory["nf0"] = 4
+        elif mu0 < theory["mt"] * thresholds["t"]:
+            theory["nf0"] = 5
+        else:
+            theory["nf0"] = 6
+
     # Setting the thresholds in the theory card to inf if necessary
-    theory.update({"kbThr":thresholds["b"], "ktThr":thresholds["t"] })
-    
+    theory.update({"kbThr": thresholds["b"], "ktThr": thresholds["t"]})
+
     # The Legacy function is able to construct a theory card for eko starting from an NNPDF theory
     legacy_class = runcards.Legacy(theory, {})
     theory_card = legacy_class.new_theory
 
     # construct operator card
     q2_grid = utils.generate_q2grid(
-        theory["Q0"],
+        mu0,
         q_fin,
         q_points,
-        {theory["mb"]: thresholds["b"], theory["mt"]: thresholds["t"]},
+        {
+            theory["mc"]: thresholds["c"],
+            theory["mb"]: thresholds["b"],
+            theory["mt"]: thresholds["t"],
+        },
+        theory["nf0"],
```

Comment:
I understand this work, so in principle is fine. 
When you call `generate_q2grid` with both Q2s given `nf0` is ignored. On contrary by default from the cli here we have Q2s to None so you actually want to use `nf0`, it's a bit confusing...

Reply:
Yes, the `nf0` is acting a bit like a secret parameter to generate the q2grid from the defaults.
I guess another option is _not_ to call `generate_q2grid` when there is nothing to generate (just a default being used).

Reply:
maybe that's more clean.

Reply:
Just to point out: `nf0` parameter is going to disappear, or better to move. Now that we are propagating evolution points from EKO, everything should be an evolution point (a point on a lane, rather than on the ladder), and also fitting scale will be an evolution point.

So `nf0` should be *always* used, and never `None` (despite being always 4, at least for the time being - but remember that is 3 for perturbative charm). You should be allowed to discard the `nf` information only at the latest possible time (i.e. when generating the LHAPDF grid file).

`nf0` is an EKO extension to the NNPDF theory database, and inferred from the comparison of `Q0` with `mc * kcThr`, but it would be better to always have it explicit, and infer as little as possible.